pylint/checkers/refactoring/refactoring_checker.py

```diff
@@ -1862,32 +1866,28 @@ def _check_unnecessary_comprehension(self, node: nodes.Comprehension) -> None:
         if expr_list == target_list and expr_list:
             args: tuple[str] | None = None
             inferred = utils.safe_infer(node.iter)
-            if isinstance(node.parent, nodes.DictComp) and isinstance(
-                inferred, astroid.objects.DictItems
-            ):
-                args = (f"dict({node.iter.func.expr.as_string()})",)
-            elif isinstance(node.parent, nodes.ListComp) and isinstance(
-                inferred, nodes.List
-            ):
-                args = (f"list({node.iter.as_string()})",)
-            elif isinstance(node.parent, nodes.SetComp) and isinstance(
-                inferred, nodes.Set
-            ):
-                args = (f"set({node.iter.as_string()})",)
+            match (node.parent, inferred):
+                case [nodes.DictComp(), astroid.objects.DictItems()]:
+                    args = (f"dict({node.iter.func.expr.as_string()})",)
+                case [nodes.ListComp(), nodes.List()]:
+                    args = (f"list({node.iter.as_string()})",)
+                case [nodes.SetComp(), nodes.Set()]:
+                    args = (f"set({node.iter.as_string()})",)
             if args:
                 self.add_message(
                     "unnecessary-comprehension", node=node.parent, args=args
                 )
                 return
 
-            if isinstance(node.parent, nodes.DictComp):
-                func = "dict"
-            elif isinstance(node.parent, nodes.ListComp):
-                func = "list"
-            elif isinstance(node.parent, nodes.SetComp):
-                func = "set"
-            else:
-                return
+            match node.parent:
+                case nodes.DictComp():
+                    func = "dict"
+                case nodes.ListComp():
+                    func = "list"
+                case nodes.SetComp():
+                    func = "set"
+                case _:  # pragma: no cover
+                    raise AssertionError
```

Comment:
```suggestion
            func = {
                nodes.DictComp(): "dict",
                nodes.ListComp(): "list",
                nodes.SetComp(): "set",
            }.get(type(node.parent), "unknown comprehension")
```
Kinda prefer this tbh ?

Reply:
Not sure this is really better. The dict is constructed just for the lookup, type checkers would likely have a hard time understanding it and it's easy to introduce bugs, e.g. the `()` for the comp classes are wrong. It can also break if we ever subclass these. There is a reason pylint has a check to suggest `isinstance` over `== type(...)`.
https://pylint.pycqa.org/en/latest/user_guide/messages/convention/unidiomatic-typecheck.html